src/state/slices/validatorDataSlice/validatorDataSlice.ts

```diff
@@ -0,0 +1,92 @@
+import { createSlice } from '@reduxjs/toolkit'
+import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
+// @ts-ignore this will fail at 'file differs in casing' error
+import { ChainAdapter as CosmosChainAdapter } from '@shapeshiftoss/chain-adapters/dist/cosmosSdk/cosmos/CosmosChainAdapter'
+import { chainAdapters } from '@shapeshiftoss/types'
+import { getChainAdapters } from 'context/PluginProvider/PluginProvider'
+
+import { cosmosChainId } from '../portfolioSlice/utils'
+
+export type PubKey = string
+
+type SingleValidatorDataArgs = { validatorAddress: PubKey }
+
+export type Validators = {
+  validators: chainAdapters.cosmos.Validator[]
+}
+
+export type ValidatorData = {
+  byValidator: ValidatorDataByPubKey
+  validatorIds: string[]
+}
+
+export type ValidatorDataByPubKey = {
+  [k: PubKey]: chainAdapters.cosmos.Validator
+}
+
+const initialState: ValidatorData = {
+  byValidator: {},
+  validatorIds: [],
+}
+
+const updateOrInsertValidatorData = (
+  validatorDataState: ValidatorData,
+  validators: chainAdapters.cosmos.Validator[],
+) => {
+  validators.forEach(validator => {
+    validatorDataState.validatorIds.push(validator.address)
+    validatorDataState.byValidator[validator.address] = validator
+  })
+}
+
+export const validatorData = createSlice({
+  name: 'validatorData',
+  initialState,
+  reducers: {
+    clear: () => initialState,
+    upsertValidatorData: (
+      validatorDataState,
+      { payload }: { payload: { validators: chainAdapters.cosmos.Validator[] } },
+    ) => {
+      updateOrInsertValidatorData(validatorDataState, payload.validators)
+    },
+  },
+})
+
+export const validatorDataApi = createApi({
+  reducerPath: 'validatorDataApi',
+  // not actually used, only used to satisfy createApi, we use a custom queryFn
+  baseQuery: fetchBaseQuery({ baseUrl: '/' }),
+  // 5 minutes caching against overfetching. The only thing that can change on new Tx is effectively TVL and APR
+  // The first won't noticeably change given the Million fiat precision we use, and the former effectively won't noticeably change either in such timeframe
+  keepUnusedDataFor: 300,
+  // refetch if network connection is dropped, useful for mobile
+  refetchOnReconnect: true,
+  endpoints: build => ({
+    getValidatorData: build.query<chainAdapters.cosmos.Validator, SingleValidatorDataArgs>({
+      queryFn: async ({ validatorAddress }, { dispatch }) => {
+        const chainAdapters = getChainAdapters()
+        const adapter = (await chainAdapters.byChainId(cosmosChainId)) as CosmosChainAdapter
+        try {
+          const data = await adapter.getValidator(validatorAddress)
+          dispatch(
+            validatorData.actions.upsertValidatorData({
+              validators: [data],
```

Comment:
is this for future compatibility? we're only fetching one?

Reply:
Yeah, it's a relic of the previous implementation which IMO makes sense to keep: even though we're now getting a single validator's data (`getValidatorData`), I'm still more confident in keeping it with the previous implementation e.g `validators.forEach(validator => immerMutateFunction)`. The forEach is a low-cost operation, and it will make it compatible with multiple validators in the future if we do need it.